test/list.go

```diff
@@ -0,0 +1,498 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package test
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/Peripli/service-manager/pkg/query"
+
+	. "github.com/onsi/ginkgo/extensions/table"
+
+	"net/http"
+
+	. "github.com/onsi/ginkgo"
+
+	"github.com/Peripli/service-manager/test/common"
+)
+
+type listOpEntry struct {
+	expectedResourcesBeforeOp []common.Object
+
+	queryTemplate              string
+	queryArgs                  common.Object
+	expectedResourcesAfterOp   []common.Object
+	unexpectedResourcesAfterOp []common.Object
+	expectedStatusCode         int
+}
+
+func DescribeListTestsFor(ctx *common.TestContext, t TestCase) bool {
+	var r []common.Object
+	var rWithMandatoryFields common.Object
+
+	attachLabel := func(obj common.Object) common.Object {
+		patchLabelsBody := make(map[string]interface{})
+		patchLabels := []query.LabelChange{
+			{
+				Operation: query.AddLabelOperation,
+				Key:       "labelKey1",
+				Values:    []string{"1"},
+			},
+			{
+				Operation: query.AddLabelOperation,
+				Key:       "labelKey2",
+				Values:    []string{"str"},
+			},
+			{
+				Operation: query.AddLabelOperation,
+				Key:       "labelKey3",
+				Values:    []string{`{"key1": "val1", "key2": "val2"}`},
+			},
+		}
+		patchLabelsBody["labels"] = patchLabels
+
+		By(fmt.Sprintf("Attempting to patch resource of %s with labels as labels are declared supported", t.API))
+		ctx.SMWithOAuth.PATCH(t.API + "/" + obj["id"].(string)).WithJSON(patchLabelsBody).
+			Expect().
+			Status(http.StatusOK)
+
+		result := ctx.SMWithOAuth.GET(t.API + "/" + obj["id"].(string)).
+			Expect().
+			Status(http.StatusOK).JSON().Object()
+		result.ContainsKey("labels")
+		r := result.Raw()
+		return r
+	}
+
+	By(fmt.Sprintf("Attempting to create a random resource of %s with mandatory fields only", t.API))
+	rWithMandatoryFields = t.RndResourceWithoutNullableFieldsBlueprint(ctx)
+	for i := 0; i < 4; i++ {
```

Comment:
4? does it have to be 4 or it could be anything?

Reply:
Just a random number greater than 2 will suffice